File: Templates/Python/range-and-or.py
from typing import List
import numpy as np

class RangeAndQueries:
    def __init__(self, arr: List[int]):
        self.MAX = 100000
        self.bitscount = 32
        self.prefix_count = np.zeros((self.bitscount, self.MAX))
        self.find_prefix_count(arr)

    def find_prefix_count(self, arr: List[int]) -> None:
        """
        Function to find the prefix count

        Args:
            arr (List[int]): Input array
        """
        n = len(arr)
        for i in range(self.bitscount):
            self.prefix_count[i][0] = ((arr[0] >> i) & 1)
            for j in range(1, n):
                self.prefix_count[i][j] = ((arr[j] >> i) & 1)
                self.prefix_count[i][j] += self.prefix_count[i][j - 1]

    def range_and(self, l: int, r: int) -> int:
        """
        Function to answer range AND query

        Args:
            l (int): Left index of the range
            r (int): Right index of the range

        Returns:
            int: Result of range AND query
        """
        ans = 2**31 - 1  # Initialize answer with all bits set to 1
        for i in range(self.bitscount):
            x = 0
            if l == 0:
                x = self.prefix_count[i][r]
            else:
                x = self.prefix_count[i][r] - self.prefix_count[i][l - 1]
            if x == r - l + 1:
                ans &= ~(1 << i)
        return ans

    def multiple_range_and_queries(self, queries: List[List[int]]) -> List[int]:
        """
        Function to answer multiple range AND queries

        Args:
            queries (List[List[int]]): List of query ranges

        Returns:
            List[int]: List of results for each query
        """

        results = []
        for query in queries:
            results.append(self.range_and(query[0], query[1]))
        return results

File: Templates/C++/range-and-or.cpp
#include <iostream>
#include <vector>

class RangeAndOrQueries {
public:
  RangeAndOrQueries(const std::vector<int> &arr) {
    MAX = 100000;
    bitscount = 32;
    prefix_count.resize(bitscount, std::vector<int>(MAX, 0));
    find_prefix_count(arr);
  }

  void find_prefix_count(const std::vector<int> &arr) {
    int n = arr.size();
    for (int i = 0; i < bitscount; ++i) {
      prefix_count[i][0] = ((arr[0] >> i) & 1);
      for (int j = 1; j < n; ++j) {
        prefix_count[i][j] = ((arr[j] >> i) & 1) + prefix_count[i][j - 1];
      }
    }
  }

  int range_and(int l, int r) {
    int ans = (1 << 31) - 1; // Initialize answer with all bits set to 1
    for (int i = 0; i < bitscount; ++i) {
      int x = 0;
      if (l == 0) {
        x = prefix_count[i][r];
      } else {
        x = prefix_count[i][r] - prefix_count[i][l - 1];
      }
      if (x != r - l + 1) {
        ans &= ~(1 << i);
      }
    }
    return ans;
  }

  std::vector<int>
  multiple_range_and_queries(const std::vector<std::vector<int>> &queries) {
    std::vector<int> results;
    for (const auto &query : queries) {
      results.push_back(range_and(query[0], query[1]));
    }
    return results;
  }

private:
  int MAX;
  int bitscount;
  std::vector<std::vector<int>> prefix_count;
};
